0001-scripting.md

```diff
@@ -0,0 +1,144 @@
+* Modifying existing component types on the fly would only work on component types that were created from the scripting environment, but I think it's a reasonable constraint. Modifying a component in dynamic state would simply be a matter of iterating over the `Vec<u8>` storage and moving the data inside around. Modifying a component in statically-collapsed state however is a bit trickier, but not that hard either. You need to replace all static instances of that component into a dynamic instance of the new component.
+* Removing an existing component type at least is easy. Just remove all instances of it and leave it all alone, scared, in a white void up to the next reboot where you end its sufferings.
+
+On the matter of resources, they simply can behave the same way component storages do.
+
+As specified, resources and components would declared using a schema language (the exact language is yet to be determined), offering the user numerous primitives such as different number types, strings, vectors, algebra primitives, etc. That way, the user can represent complex data structures from the schema. All those primitives must be sufficiently represented with C types so that they can be shared by all drivers. One important aspect is that, when possible, they do not hold references.
+
+A user can potentially force a reference to be stored in a component of a resource. But this is out of the scope of the scripting support: local safety has already been traded off by not using Rust. We only need to ensure the scripting languages can not disrupt the safety of code outside of the scripting world.
+
+## Satisfaction of goals
+
+Let us now take every goals one by one and explain how this proposal solves them specifically.
+
+* Support any language: a language driver can be written for any programming language that supports C FFI, which corresponds to any serious programming language available.
+* Enforce the safety of the ECS: the rules enforced by a language driver's runner part will guarantee that the ECS is safe.
+* Close to no overhead: all the communication happens through simple C FFI without additional processing, which is the fastest possible solution.
+* Expandable: language drivers do not require modifying the engine and are very modular, making them easy to plug in.
+* Little maintenance for each languages: as the augmented FFI is automatically discovered at build time, the language drivers expand the API they expose by themselves, without a need for maintenance.
+* Support the creation of components and resources: the primitives exposed in the schema will offer a large palette to design complex structures.
+* Support hotswapping: see above.
+* Independent from Rust: if enough features are available to scripts, it will be possible to make it no longer necessary to write any Rust code to use the engine to its full potential.
+
+## Limitations and drawbacks
```

Comment:
What happens if a script, say, goes into an infinite loop? Would we try to detect that on the Rust side? Or leave it to the user to debug?

Reply:
If a script does not return, then you would get the same behavior as with a Rust native system. Unless we want to improve the situation on the Rust native side, I don't really see what else we could do.

Reply:
FWIW: One of the complaints I see pop up in Unity-land over and over is that an accidental infinite-loop in a MonoBehaviour completely freezes the editor, and you have to restart Unity and hope you didn't lose any work.

So long as the tooling avoids this particular pitfall, though, having an infinite-loop script do unimaginable things to the running system is perfectly fine.

Reply:
At some point, I still think it might be worth some thought on ways to help control runaway scripts. Something like Erlang does with NIFs where it requires them to report reductions, maybe.